walkeeper/src/replication.rs

```diff
@@ -0,0 +1,238 @@
+        loop {
+            match FeMessage::read_from(&mut stream_in)? {
+                FeMessage::CopyData(m) => {
+                    let feedback = HotStandbyFeedback::des(&m.body)?;
+                    timeline.add_hs_feedback(feedback)
+                }
+                msg => {
+                    info!("unexpected message {:?}", msg);
+                }
+            }
+        }
+    }
+
+    /// Helper function that parses a pair of LSNs.
+    fn parse_start_stop(cmd: &[u8]) -> Result<(Lsn, Lsn)> {
+        let re = Regex::new(r"([[:xdigit:]]+/[[:xdigit:]]+)").unwrap();
+        let caps = re.captures_iter(str::from_utf8(&cmd[..])?);
+        let mut lsns = caps.map(|cap| cap[1].parse::<Lsn>());
+        let start_pos = lsns
+            .next()
+            .ok_or_else(|| anyhow!("failed to find start LSN"))??;
+        let stop_pos = lsns.next().transpose()?.unwrap_or(Lsn(0));
+        Ok((start_pos, stop_pos))
+    }
+
+    /// Helper function for opening a wal file.
+    fn open_wal_file(wal_file_path: &Path) -> Result<File> {
+        // First try to open the .partial file.
+        let mut partial_path = wal_file_path.to_owned();
+        partial_path.set_extension("partial");
+        if let Ok(opened_file) = File::open(&partial_path) {
+            return Ok(opened_file);
+        }
+
+        // If that failed, try it without the .partial extension.
+        match File::open(&wal_file_path) {
+            Ok(opened_file) => return Ok(opened_file),
+            Err(e) => {
+                error!("Failed to open log file {:?}: {}", &wal_file_path, e);
+                return Err(e.into());
+            }
+        }
+    }
+
+    ///
+    /// Handle START_REPLICATION replication command
+    ///
+    pub fn run(&mut self, cmd: &Bytes) -> Result<()> {
+        // spawn the background thread which receives HotStandbyFeedback messages.
+        let bg_timeline = Arc::clone(self.timeline.get());
+        let bg_stream_in = self.stream_in.take().unwrap();
+
+        thread::spawn(move || {
+            if let Err(err) = Self::background_thread(bg_stream_in, bg_timeline) {
+                error!("socket error: {}", err);
+            }
+        });
+
+        let (mut start_pos, mut stop_pos) = Self::parse_start_stop(&cmd)?;
+
+        let wal_seg_size = self.timeline.get().get_info().server.wal_seg_size as usize;
+        if wal_seg_size == 0 {
+            bail!("Can not start replication before connecting to wal_proposer");
+        }
+        let (wal_end, timeline) = self.timeline.find_end_of_wal(&self.conf.data_dir, false);
+        if start_pos == Lsn(0) {
+            start_pos = wal_end;
+        }
+        if stop_pos == Lsn(0) && self.appname == Some("wal_proposer_recovery".to_string()) {
+            stop_pos = wal_end;
+        }
+        info!("Start replication from {} till {}", start_pos, stop_pos);
+
+        let mut outbuf = BytesMut::new();
+        BeMessage::write(&mut outbuf, &BeMessage::Copy);
```

Comment:
According to docs, you must reserve enough space in outbuf manually beforehand. Strangely, it doesn't fail and I see capacity() normally grows as buffer is filled...

Reply:
`BeMessage::write` accepts a `&mut BytesMut`, and calls `put_u8`, etc. on it. That's part of the [`BufMut`](https://docs.rs/bytes/1.0.1/bytes/buf/trait.BufMut.html) trait, and it writes to the end, extending the internal buffer as needed.

Some of the `BytesMut` docs seem a little misleading, since they call `reserve` before writing to the buffer, which is not needed.

This test passes:
```
#[test]
fn test_bytesmut() {
    let mut buf = BytesMut::new();
    assert_eq!(buf.len(), 0);
    assert_eq!(buf.capacity(), 0);
    buf.put_u8(b'A');
    buf.put_u32(0x01020304);
    assert_eq!(buf.len(), 5);
    assert_eq!(buf.as_ref(), &[65, 1, 2, 3, 4]);
}
```